contributors/design-proposals/storage/volume-snapshotting.md

```diff
@@ -21,498 +25,529 @@ For the first version of snapshotting support in Kubernetes, only on-demand snap
+
+        // Representes the lates available observations about the volume snapshot
+        Conditions []VolumeSnapshotDataCondition `json:"conditions" protobuf:"bytes,2,rep,name=conditions"`
+}
+
+// VolumeSnapshotDataConditionType is the type of the VolumeSnapshotData condition
+type VolumeSnapshotDataConditionType string
+
+// These are valid conditions of a volume snapshot.
+const (
+        // VolumeSnapshotDataReady is added when the on-disk snapshot has been successfully created.
+        VolumeSnapshotDataConditionReady VolumeSnapshotDataConditionType = "Ready"
+        // VolumeSnapshotDataPending is added when the on-disk snapshot has been successfully created but is not available to use.
+        VolumeSnapshotDataConditionPending VolumeSnapshotDataConditionType = "Pending"        // VolumeSnapshotDataError is added but the on-disk snapshot is failed to created
+        VolumeSnapshotDataConditionError VolumeSnapshotDataConditionType = "Error"
+)
+
+// VolumeSnapshotDataCondition describes the state of a volume snapshot  at a certain point.
+type VolumeSnapshotDataCondition struct {
+        // Type of volume snapshot condition.
+        Type VolumeSnapshotDataConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=VolumeSnapshotDataConditionType"`
+        // Status of the condition, one of True, False, Unknown.
+        Status core_v1.ConditionStatus `json:"status" protobuf:"bytes,2,opt,name=status,casttype=ConditionStatus"`
+        // The last time the condition transitioned from one status to another.
+        // +optional
+        LastTransitionTime metav1.Time `json:"lastTransitionTime" protobuf:"bytes,3,opt,name=lastTransitionTime"`
+        // The reason for the condition's last transition.
+        // +optional
+        Reason string `json:"reason" protobuf:"bytes,4,opt,name=reason"`
+        // A human readable message indicating details about the transition.
+        // +optional
+        Message string `json:"message" protobuf:"bytes,5,opt,name=message"`
+}
+
+// Represents the actual location and type of the snapshot. Only one of its members may be specified.
+type VolumeSnapshotDataSource struct {
+	// HostPath represents a directory on the host.
+	// Provisioned by a developer or tester.
+	// This is useful for single-node development and testing only!
+	// On-host storage is not supported in any way and WILL NOT WORK in a multi-node cluster.
+	// More info: https://kubernetes.io/docs/concepts/storage/volumes#hostpath
+	// +optional
+	HostPath *HostPathVolumeSnapshotSource `json:"hostPath,omitempty"`
+	// AWSElasticBlockStore represents an AWS Disk resource that is attached to a
+	// kubelet's host machine and then exposed to the pod.
+	// More info: https://kubernetes.io/docs/concepts/storage/volumes#awselasticblockstore
+	// +optional
+	AWSElasticBlockStore *AWSElasticBlockStoreVolumeSnapshotSource `json:"awsElasticBlockStore,omitempty"`
```

Comment:
I was expecting the cloud provider-related API and controller to be out of tree. Isn't the ideal scenario that any storage provider would be able to support snapshotting their storage, interacting with the core kubernetes API only via VolumeSnapshot/VolumeSnapshotData objects?

Reply:
to that end, it would probably make sense for there to be a place for opaque, storage-provider-specific parameters in the VolumeSnapshotData, so the storage provider can persist the info required to locate the snapshot (similar to the parameters in StorageClass and FlexVolumeSource)

Reply:
Thanks @liggitt for the comments.  This design is modeled after the design of PersistentVolume and PersistentVolumeClaim. For example, PersistentVolumeSource contains storage provider data as follows:

type PersistentVolumeSource struct {
// GCEPersistentDisk represents a GCE Disk resource that is attached to a
// kubelet's host machine and then exposed to the pod.
// +optional
GCEPersistentDisk *GCEPersistentDiskVolumeSource

// AWSElasticBlockStore represents an AWS EBS disk that is attached to a
// kubelet's host machine and then exposed to the pod
// +optional
AWSElasticBlockStore *AWSElasticBlockStoreVolumeSource
...
}

Having opaque, storage-provider-specific parameters in the VolumeSnapshotData is an option we can discuss about.